server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java

```diff
@@ -0,0 +1,357 @@
+        PerRecipientHeaders.Header header = PerRecipientHeaders.Header.builder()
+            .name("any")
+            .value("any")
+            .build();
+        getMailQueue().enQueue(FakeMail.builder()
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .name(NAME1)
+            .addHeaderForRecipient(header, MailAddressFixture.ANY_AT_JAMES)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getPerRecipientSpecificHeaders()
+            .getHeadersForRecipient(MailAddressFixture.ANY_AT_JAMES))
+            .containsOnly(header);
+    }
+
+    @Test
+    default void dequeueShouldBeFifo() throws Exception {
+        getMailQueue().enQueue(FakeMail.builder()
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .name(NAME1)
+            .build());
+        getMailQueue().enQueue(FakeMail.builder()
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .name(NAME2)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo(NAME1);
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo(NAME2);
+    }
+
+    @Test
+    default void dequeueCouldBeInterleaving() throws Exception {
+        getMailQueue().enQueue(FakeMail.builder()
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .name(NAME1)
+            .build());
+        getMailQueue().enQueue(FakeMail.builder()
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .name(NAME2)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        mailQueueItem2.done(true);
```

Comment:
what happens if you switch the 2 done calls?

Reply:
I will add the test.